src/rgw/rgw_sal_d3n.cc

```diff
@@ -0,0 +1,324 @@
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation. See file COPYING.
+ *
+ */
+
+#include "rgw_sal.h"
+#include "rgw_sal_rados.h"
+#include "rgw_sal_d3n.h"
+#include "rgw_aio_throttle.h"
+
+#define dout_subsys ceph_subsys_rgw
+#define dout_context g_ceph_context
+
+namespace rgw { namespace sal {
+
+class RadosStore;
+
+int D3NFilterDriver::initialize(CephContext *cct, const DoutPrefixProvider *dpp)
+{
+  FilterDriver::initialize(cct, dpp);
+  d3n_cache->init(cct);
+
+  return 0;
+}
+
+std::unique_ptr<Object> D3NFilterDriver::get_object(const rgw_obj_key& k)
+{
+  std::unique_ptr<Object> obj = next->get_object(k);
+  return std::make_unique<D3NFilterObject>(std::move(obj), this);
+}
+
+std::unique_ptr<Bucket> D3NFilterDriver::get_bucket(const RGWBucketInfo& i)
+{
+  std::unique_ptr<Bucket> bp = next->get_bucket(i);
+  return std::make_unique<D3NFilterBucket>(std::move(bp), this);
+}
+
+std::unique_ptr<Object> D3NFilterBucket::get_object(const rgw_obj_key& k)
+{
+  std::unique_ptr<Object> obj = next->get_object(k);
+  return std::make_unique<D3NFilterObject>(std::move(obj), this, this->filter);
+}
+
+std::unique_ptr<Object::ReadOp> D3NFilterObject::get_read_op()
+{
+  std::unique_ptr<Object::ReadOp> rop = next->get_read_op();
+  return std::make_unique<D3NFilterReadOp>(std::move(rop), this, filter);
+}
+
+int D3NFilterObject::D3NFilterReadOp::prepare(optional_yield y, const DoutPrefixProvider* dpp)
+{
+  next->params.mod_ptr = params.mod_ptr;
+  next->params.unmod_ptr = params.unmod_ptr;
+  next->params.high_precision_time = params.high_precision_time;
+  next->params.mod_zone_id = params.mod_zone_id;
+  next->params.mod_pg_ver = params.mod_pg_ver;
+  next->params.if_match = params.if_match;
+  next->params.if_nomatch = params.if_nomatch;
+  next->params.lastmod = params.lastmod;
+
+  auto ret = next->prepare(y, dpp);
+  if (ret < 0) {
+    return ret;
+  }
+
+  if (! this->source->have_instance()) {
+    RGWObjState* state = nullptr;
+    if (this->source->get_obj_state(dpp, &state, y) == 0) {
+      auto it = state->attrset.find(RGW_ATTR_ID_TAG);
+      if (it != state->attrset.end()) {
+        bufferlist bl = it->second;
+        this->source->set_object_version(bl.c_str());
+        ldpp_dout(dpp, 20) << __func__ << "id tag version is: " << this->source->get_object_version() << dendl;
+      } else {
+        ldpp_dout(dpp, 20) << __func__ << "Failed to find id tag" << dendl;
+      }
+    }
+  }
+
+  return 0;
+}
+
+void D3NFilterObject::D3NFilterReadOp::cancel() {
+  aio->drain();
+}
+
+int D3NFilterObject::D3NFilterReadOp::drain(const DoutPrefixProvider* dpp) {
+  auto c = aio->wait();
+  while (!c.empty()) {
+    int r = flush(dpp, std::move(c));
+    if (r < 0) {
+      cancel();
+      return r;
+    }
+    c = aio->wait();
+  }
+
+  c = aio->drain();
+  int r = flush(dpp, std::move(c));
```

Comment:
if we're calling `drain()`, we shouldn't need to loop over calls to `wait()` first

Reply:
@cbodley : What is the best way to ensure that `pending list` is empty when destructor of aio is invoked?

Reply:
first stop submitting new ios, then call drain to `// wait for all outstanding completions and return their results`